Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,130 @@
+
+3. Or even measuring samples queried per second for Prometheus instances under heavy query load.
+
+- Prometheus Servers: `rate(prometheus_engine_query_samples_total[5m])`
+
+Modern Custom Horizontal Pod Autoscalers (HPAs), such as Keda are already capable of [scaling CRDs](https://keda.sh/docs/2.11/concepts/scaling-deployments/#scaling-of-custom-resources) based on different parameters, such as [CPU, Memory](https://keda.sh/docs/2.11/scalers/cpu/) or even [custom Prometheus metrics](https://keda.sh/docs/2.11/scalers/prometheus/).
+
+### Scale Status subresource
+
+When working with CRDs, common HPAs (e.g. Keda) depends on a Status subresource called [Scale](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource). Unfortunately, Prometheus-Operator resources still don't implement this field. There was an attempt to implement this in https://github.com/prometheus-operator/prometheus-operator/pull/4735, but was reverted because the PR was scaling replicas instead of shards. Although scaling replicas can help with High-Availability, it is expensive and hard to manage since it duplicates scrapes while not reducing the load on top of Prometheus because all replicas still scrape the same targets. Sharding serves as a better autoscaling strategy since the load is spread into all instances instead of getting duplicated.
+
+With only this change, Prometheus Agents can already be horizontally scaled without problems, but for Prometheus Servers it gets a little more complicated.
+
+### Graceful scale down of Prometheus Servers
+
+Prometheus Servers don't behave like the usual stateless applications. A few cases require Prometheus pods to still be available even after a scale down event is triggered:
+
+* Thanos proxying queries back to the sidecars to retrieve data that still wasn't uploaded to object storage.
+* Prometheus servers with long retention periods still hold data that users might want to query.
+
+If we simply delete Prometheus pods during scale-down, data loss will be a common problem, so a more complex scale-down strategy needs to be used here.
+
+We propose that Prometheus-Operator introduces a new field to the Prometheus Server CRDs:
+
+```yaml
+apiVersion: monitoring.coreos.com
+kind: Prometheus
+metadata:
+  name: example
+spec:
+  autoscaling:
+    scaleDownStrategy: force|preventDataLoss # Default: force
+```
+
+If scaleDownStrategy == force, the operator would simply delete the pod on the spot, not caring about the data stored in that particular shard. Although not the safest approach, it is exactly the current behavior nowadays and changing it might surprise a lot of users.
+
+If scaleDownStrategy == preventDataLoss then Prometheus-Operator would first mark the pod with an annotation:
+
+```yaml
+prometheus.operator.io/safe-deletion-timestamp: X
+```
+
+Where X is the timestamp of the current retention period.
```

Comment:
If we have a sidecar shipping blocks away, do we want to wait for the prometheus retention period still or do we want to wait for the youngest block at the time of the scale down to be confirmed shipped?

Note: such a feature could be a later improvement if needed, not needed for the first version imo

Reply:
Having the Sidecar uploading the blocks created by the snapshot feature, we can shutdown the Prometheus straight the way and we don't need to wait the deletion period.

Since it requires a implementation on Thanos I also think this is improving version

Reply:
I like @simonpasquier's idea of having a default value (retention period) but allowing users to customize the retention to different periods, e.g. 2h.

Also mentioned by simon, I believe we can't assume it is okay to delete after blocks are uploaded because the user can still depend on that data for alerting purposes

Reply:
> Also mentioned by simon, I believe we can't assume it is okay to delete after blocks are uploaded because the user can still depend on that data for alerting purposes

Good point :+1: